packages/firestore/test/integration/util/composite_index_test_helper.ts

```diff
@@ -0,0 +1,175 @@
+  }
+
+  // Adds test-specific fields to a document, including the testId and expiration date.
+  addTestSpecificFieldsToDoc(doc: DocumentData): DocumentData {
+    return {
+      ...doc,
+      [this.TEST_ID_FIELD]: this.testId,
+      expireAt: new Timestamp( // Expire test data after 24 hours
+        Timestamp.now().seconds + 24 * 60 * 60,
+        Timestamp.now().nanoseconds
+      )
+    };
+  }
+
+  // Helper method to hash document keys and add test-specific fields for the provided documents.
+  private prepareTestDocuments(docs: { [key: string]: DocumentData }): {
+    [key: string]: DocumentData;
+  } {
+    const result: { [key: string]: DocumentData } = {};
+    for (const key in docs) {
+      if (docs.hasOwnProperty(key)) {
+        result[this.toHashedId(key)] = this.addTestSpecificFieldsToDoc(
+          docs[key]
+        );
+      }
+    }
+    return result;
+  }
+
+  // Asserts that the result of running the query while online (against the backend/emulator) is
+  // the same as running it while offline. The expected document Ids are hashed to match the
+  // actual document IDs created by the test helper.
+  async assertOnlineAndOfflineResultsMatch(
+    query: Query,
+    ...expectedDocs: string[]
+  ): Promise<void> {
+    return checkOnlineAndOfflineResultsMatch(
+      query,
+      ...this.toHashedIds(expectedDocs)
+    );
+  }
+
+  // Adds a filter on test id for a query.
+  query<AppModelType, DbModelType extends DocumentData>(
+    query_: Query<AppModelType, DbModelType>,
+    ...queryConstraints: QueryConstraint[]
+  ): Query<AppModelType, DbModelType> {
+    return internalQuery(
+      query_,
+      where(this.TEST_ID_FIELD, '==', this.testId),
+      ...queryConstraints
+    );
+  }
+
+  // Adds a filter on test id for a composite query.
+  compositeQuery<AppModelType, DbModelType extends DocumentData>(
+    query_: Query<AppModelType, DbModelType>,
+    compositeFilter: QueryCompositeFilterConstraint,
+    ...queryConstraints: QueryNonFilterConstraint[]
+  ): Query<AppModelType, DbModelType> {
+    return internalQuery(
+      query_,
+      and(where(this.TEST_ID_FIELD, '==', this.testId), compositeFilter),
+      ...queryConstraints
+    );
+  }
+
+  // Adds a document to a Firestore collection with test-specific fields.
+  addDoc<T, DbModelType extends DocumentData>(
+    reference: CollectionReference<T, DbModelType>,
+    data: object
+  ): Promise<DocumentReference<T, DbModelType>> {
+    const processedData = this.addTestSpecificFieldsToDoc(
+      data
+    ) as WithFieldValue<T>;
+    return addDocument(reference, processedData);
+  }
+
+  // Sets a document in Firestore with test-specific fields.
+  setDoc<T, DbModelType extends DocumentData>(
+    reference: DocumentReference<T, DbModelType>,
+    data: object
+  ): Promise<void> {
+    const processedData = this.addTestSpecificFieldsToDoc(
+      data
+    ) as WithFieldValue<T>;
+    return setDocument(reference, processedData);
+  }
```

Comment:
I know these functions aren't needed right now, but we should probably implement them to have a complete set of utility functions in this class. (_note: this is rough untested code._)

```js
// Updates a document in Firestore.
// This is is the same as making the update on the doc directly with merge=true.
updateDoc<T, DbModelType extends DocumentData>(
    reference: DocumentReference<T, DbModelType>,
    updates: object
  ): Promise<void> {
  return updateDoc(reference, updates);
}

// Gets the document content.
async getDoc<AppModelType, DbModelType extends DocumentData>(
    reference: DocumentReference<AppModelType, DbModelType>
): Promise<DocumentSnapshot<AppModelType, DbModelType>> {
    let docSnapshot = await getDocument<AppModelType, DbModelType>(reference);
    docSnapshot._document?.data?.delete(field(this.TTL_FIELD));
    docSnapshot._document?.data?.delete(field(this.TEST_ID_FIELD));
    return docSnapshot;
}

getDocs<AppModelType, DbModelType extends DocumentData>(
  query: Query<AppModelType, DbModelType>
): Promise<QuerySnapshot<AppModelType, DbModelType>> {
  // querySnapshot = await getDocs(query);
  // iterate over documents in the querySnapshot, and do the same thing as above.
  // should probably make a function named "removeTestSpecificFieldsFromDoc" and
  // and call that in these two functions.
}
```